### PR TITLE
Support incomplete insertor

### PR DESCRIPTION
**Pull request by karenzheung**
_Pull requests into Flux require the following. Submitter should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:. Reviewers should complete every item on the bulleted list when reviewing._

## Submitter:

You can now insert and inserter as incomplete. Typing @cond... and selecting from the autoreplace will replace it with an incomplete @condition inserter. You can also type @condition then press space, which will also replace it with incomplete. 

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [ ] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [ ] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [x] Branch implements task as expected
- [x] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- [x] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [x] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [x] Document any new APIs/extension points
- [x] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [x] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- [x] Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

**Comment by mgramigna** [comment]
This looks great! Two comments for now:

[MG1] We can remove this commented out code https://github.com/FluxNotes/flux/blob/36cc71f1ea7ddeb38c7f1fa0b8abfa5fb0d90039/src/notes/FluxNotesEditor.scss#L154-L175

[MG2] If I type `@condition` into the editor, it renders as incomplete, and then hit `Esc` so I close the portal, I can't get back to the picker menu for that incomplete shortcut.

I'm not sure what the best way to handle MG2 is. We don't really encounter this problem on master because even after you've selected "condition" as the shortcut from the menu, it is not inserted into the editor until you select the value. Approaches could be:

1) Get the portal for selecting the value to re-open when the cursor is next to `@condition`
2) If they close the portal with `Esc` without selecting a value, also remove the incomplete `@condition` shortcut

My guess is that option 1 is more desirable, but naturally it is probably more difficult.

**Comment by gregquinn2001** [comment]
For MG2, option 1 is required and makes it work the same as #status and other #label child creators.

**Comment by mgramigna** [comment]
[MG3] We should move lines 423 and 424 inside the if block above.  For my task, we will be reusing this functionality for pickers for the attribute shortcuts. We don't want to add these attribute shortcuts to context, so we need these lines to only happen if the shortcut is a Context

https://github.com/FluxNotes/flux/blob/36cc71f1ea7ddeb38c7f1fa0b8abfa5fb0d90039/src/notes/FluxNotesEditor.jsx#L420-L424

**Comment by karenzheung** [comment]
Fixed MG1 and MG3. Subtask ASCODCP-1846 captures MG2 and is currently in progress.

File: src/notes/FluxNotesEditor.scss [hunk 1]
```diff
@@ -108,10 +108,20 @@
...
+                &-incomplete {
+                    color: $state;
+                    display: inline;
+                    border-bottom: 1.5px dashed $missing-data-red;
+                }
             }
         
             &-creator {
-                border-bottom: 1px solid $shr-context-line;
+                border-bottom: 1px dashed $shr-context-line;
```
Comment (mgramigna) [inline src/notes/FluxNotesEditor.scss]: [MG4] When we rearranged the sass we lost consistency in the border thickness so this should be 1.5px instead of 1px

**Comment by karenzheung** [comment]
Fixed MG4

File: src/notes/FluxNotesEditor.jsx [hunk 2]
```diff
@@ -401,21 +412,19 @@ class FluxNotesEditor extends React.Component {
...
-        if (this.state.needToDelete) {
-            transform = this.suggestionDeleteExistingTransform(null, shortcut.getPrefixCharacter());
-        } else {
-            transform = this.state.state.transform();
-        }
-
-        return this.insertStructuredFieldTransform(transform, shortcut).collapseToStartOfNextText().focus().apply();
+        transform = this.resetShortcutData(shortcut, transform);
+        return transform.apply();
+       // return this.insertStructuredFieldTransform(transform, shortcut).collapseToStartOfNextText().focus().apply();
```
Comment (Dtphelan1) [inline src/notes/FluxNotesEditor.jsx]: [DP1] This should be removed if we don't need it anymore

File: src/notes/FluxNotesEditor.jsx [hunk 3]
```diff
@@ -302,13 +308,21 @@ class FluxNotesEditor extends React.Component {
...
             return this.insertPlainText(transform, shortcutTrigger);
         }
-
         let shortcut = this.props.newCurrentShortcut(shortcutC, shortcutTrigger, text, updatePatient, source);
+        if(text === "" && shortcut.needToSelectValueFromMultipleOptions()){
+            shortcut.setText(shortcutTrigger);
+        }
         shortcut.initialContextPosition = initialContextPosition;
+        
+        this.insertStructuredFieldTransform(transform, shortcut).collapseToStartOfNextText().apply();
```
Comment (Dtphelan1) [inline src/notes/FluxNotesEditor.jsx]: [DP2] We shouldn't be applying our transform if we are going to continue modifying that transform later on, as we conditionally do on line 323.

**Comment by Dtphelan1** [comment]
Also was MG2 ever fixed? Can't immediately tell from the comments.

**Comment by Dtphelan1** [comment]
[DP3] When I choose to insert a template, inserters that require values to be selected don't show up with the red underline. @oncohist does, which is good, but @condition and @medication do not. 
image

**Comment by mgramigna** [comment]
For DP3, I think ASCODCP-1907 covers template insertion, including proper notions of incomplete @ shortcuts being inserted.

For MG2, we decided this is encompassed by ASCODCP-1908

**Comment by Dtphelan1** [comment]
Thanks for the confirmation on MG2; all good on DP3 (I saw partial support and wasn't sure if we were aiming for full support here, so thanks for the clarification).

Then my comments should boil down to those few comments.

**Comment by Dtphelan1** [comment]
Addressed DP1/2

**Comment by Dtphelan1** [comment]
Aware that my changes are causing tests to fail. Investigating now.

**Comment by Dtphelan1** [comment]
Alright I think this looks good to me. Merging...
